File: lib/src/player/playback_state.dart
import 'package:music_player/music_player.dart';
import 'package:system_clock/system_clock.dart';

class PlaybackState {
  final PlayerState state;

  final int position;

  final int bufferedPosition;

  final double speed;

  final PlaybackError? error;

  final int updateTime;

  ///
  /// Get the real position in current time stamp.
  ///
  /// discuss: https://github.com/boyan01/flutter-music-player/issues/1
  ///
  int get computedPosition {
    var append = state == PlayerState.playing
        ? (SystemClock.uptime().inMilliseconds - updateTime)
        : 0;
    append = (append * speed).toInt();
    return position + append;
  }

  const PlaybackState({
    required this.state,
    required this.position,
    required this.bufferedPosition,
    required this.speed,
    this.error,
    required this.updateTime,
  });

  const PlaybackState.none()
      : this(
          state: PlayerState.none,
          position: 0,
          bufferedPosition: 0,
          speed: 1,
          updateTime: 0,
        );

  @override
  String toString() {
    return 'PlaybackState{state: $state, position: $position, bufferedPosition: $bufferedPosition, speed: $speed, error: $error, updateTime: $updateTime}';
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is PlaybackState &&
          runtimeType == other.runtimeType &&
          state == other.state &&
          position == other.position &&
          bufferedPosition == other.bufferedPosition &&
          speed == other.speed &&
          error == other.error &&
          updateTime == other.updateTime;

  @override
  int get hashCode =>
      state.hashCode ^
      position.hashCode ^
      bufferedPosition.hashCode ^
      speed.hashCode ^
      error.hashCode ^
      updateTime.hashCode;
}

enum PlayerState {
  none,
  paused,
  playing,
  buffering,
  error,
}

File: ios/Classes/PlaybackState.swift
//
// Created by yangbin on 2020/2/22.
//

import Foundation

struct PlaybackState {
    let state: State
    let position: TimeInterval
    let bufferedPosition: TimeInterval
    let speed: Float
    let error: PlaybackError?
    // duration in mills.
    let updateTime: Int
}

extension PlaybackState {
    func toMap() -> [String: Any?] {
        [
            "state": state.rawValue,
            "position": Int(position * 1000),
            "bufferedPosition": Int(bufferedPosition * 1000),
            "speed": speed,
            "error": error?.toMap(),
            "updateTime": updateTime,
        ]
    }
}

enum State: Int {
    case none = 0, paused, playing, buffering, error
}

enum ErrorType: Int {
    // detail see lib/src/playback_error.dart ErrorType.*
    case source = 0, render, unknown
}

struct PlaybackError {
    let type: ErrorType
    let message: String
}

extension PlaybackError {
    func toMap() -> [String: Any?] {
        [
            "type": type.rawValue,
            "message": message,
        ]
    }
}
